### PR TITLE
Add more metrics and clean up old code

### PR DESCRIPTION
**Pull request by EllAchE**

### CODE REVIEW

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      - name: Install Node
+        uses: actions/setup-node@v3
+        with:
+          node-version: 20
+
+      - name: Install Dependencies
+        run: npm install
+
+      - name: Run tests
+        run: npm run test
```
Comment (bennyrubanov) [inline .github/workflows/node.js.yml:25]: did you create this in Github actions UI or in VSCode?

Comment (EllAchE) [reply .github/workflows/node.js.yml:25]: it's just a yaml file. I wrote it

File: src/metrics/distances.ts [hunk 2]
```diff
@@ -1,190 +1,156 @@
...
 
-  return {
-    maxDistancePiece,
-    maxDistance,
-    distanceMap,
-    singleGameDistanceTotal,
-  };
-}
+  logResults(): void {
+    console.timeEnd('Task 2: getMoveDistanceSetOfGames');
```
Comment (bennyrubanov) [inline src/metrics/distances.ts:64]: this needs to be changed as this function name is obsolete?

File: src/metrics/distances.ts [hunk 3]
```diff
@@ -1,190 +1,156 @@
...
+import { Metric } from './metric';
+
+export class MoveDistanceMetric implements Metric {
+  distanceMap: UAPMap<{ total: number; maxSingleGame: number }>;
+  pieceWithHighestAvg: UASymbol;
+  pieceWithLowestAvg: UASymbol;
+  maxAvgDistance: number;
+  minAvgDistance: number;
+  totalDistance: number;
+  maxSingleGameTotal: number;
```
Comment (bennyrubanov) [inline src/metrics/distances.ts:13]: should be maxSingleGameDistance no?

File: src/metrics/distances.ts [hunk 4]
```diff
@@ -1,190 +1,156 @@
...
-      gameLinkWithFurthestCollectiveDistance = game.metadata
-        .find((item) => item.startsWith('[Site "'))
-        ?.replace('[Site "', '')
-        .replace('"]', '');
-    }
+    // add the single game aggregates to the state object
+    for (const uas of Object.keys(singleGameMap)) {
+      this.distanceMap[uas].total += singleGameMap[uas].total;
+      if (
+        singleGameMap[uas].total > this.distanceMap[uas].maxSingleGameDistance
```
Comment (bennyrubanov) [inline src/metrics/distances.ts:139]: where does maxSingleGameDistance come from? you have defined maxSingleGame, and maxSingleGameTotal, but not maxSingleGameDistance. i think you might be using both maxSingleGameTotal and maxSingleGameDistance for the same purpose, need to pick one

File: src/metrics/moves.ts [hunk 5]
```diff
@@ -1,164 +1,101 @@
...
+    this.link = undefined;
+    this.numMoves = 0;
+  }
 
-  // moves facts
-  console.log('MOVES FACTS:');
-  console.log(`The game with the most moves played: ${gameWithMostMoves}`);
-  console.log(`The number of moves played in that game: ${maxNumMoves}`);
+  processGame(game: { move: PrettyMove; board: Piece[] }[], gameLink?: string) {
+    if (game.length > this.numMoves) {
```
Comment (bennyrubanov) [inline src/metrics/moves.ts:20]: does game.length return the appropriate value now? i had trouble with this in the past. wouldn't it just be 3 given there are three sub arrays in the game array?

File: src/metrics/promotions.ts [hunk 6]
```diff
@@ -1,54 +1,57 @@
...
+import {
+  Pawns,
+  Piece,
+  PrettyMove,
+  PromotablePiece,
+} from '../../cjsmin/src/chess';
+import { Metric } from './metric';
+
+export class PromotionMetric implements Metric {
+  //uas pawns mapped to count of promotions to each piece type
```
Comment (bennyrubanov) [inline src/metrics/promotions.ts:10]: 3d array for promotionMap? nice. but why necessary?

File: src/metrics/promotions.ts [hunk 7]
```diff
@@ -1,54 +1,57 @@
...
-          ambigPiecePromotedToMap[piecePromotedTo] = 0;
-        }
-        ambigPiecePromotedToMap[piecePromotedTo]++;
-
-        // update promotingPieceMap
-        if (!promotingPieceMap[promotingPiece]) {
-          promotingPieceMap[promotingPiece] = 0;
-        }
-        promotingPieceMap[promotingPiece]++;
+        this.promotionMap[move.uas][move.promotion]++;
```
Comment (bennyrubanov) [inline src/metrics/promotions.ts:36]: this line: is promotion a tracked property of the moves array from cjsmin?

File: src/types.ts [hunk 8]
```diff
@@ -1,8 +1,16 @@
-import { PieceType, Square } from '../cjsmin/src/chess';
+import { Square, UASymbol } from '../cjsmin/src/chess';
 
 export interface FileReaderGame {
```
Comment (bennyrubanov) [inline src/types.ts:3]: assuming that FileReaderGame still used by GameChunks

File: src/types.ts [hunk 9]
```diff
@@ -1,8 +1,16 @@
...
   moves: string;
   metadata: string[];
 }
 
+export type UAPMap<T> = {
+  [key in UASymbol]: T;
```
Comment (bennyrubanov) [inline src/types.ts:9]: T as the key? to set it to something that wouldn't be used as starting value?

File: src/utils.ts [hunk 10]
```diff
@@ -6,3 +15,61 @@ export function orderObject(unordered) {
...
+        revengeKills: 0,
+      };
+    }
+  }
+
+  return squareCaptureInfo;
+}
+
+/**
+ * A utitily function to create an object with unambiguous piece symbols as keys
```
Comment (bennyrubanov) [inline src/utils.ts:50]: utility spelled wrong ;)

File: src/utils.ts [hunk 11]
```diff
@@ -6,3 +15,61 @@ export function orderObject(unordered) {
...
+ */
+export function createUAPMap<T extends Object>(object: T): UAPMap<T> {
+  const map = {};
+  for (const uap of ALL_UNAMBIGUOUS_PIECE_SYMBOLS) {
+    map[uap] = { ...object };
+  }
+  return map as { [key in UASymbol]: T };
+}
+
+export function createBoardMap<T extends Object>(object: T): BoardMap<T> {
```
Comment (bennyrubanov) [inline src/utils.ts:60]: in what contexts do you imagine these being used?

File: tests/enPassant.test.ts [hunk 12]
```diff
@@ -3,7 +3,7 @@ import { Chess } from '../cjsmin/src/chess';
 
 const pgnEP = `1. d2d4 f7f5 2. b2b3 e7e6 3. c1b2 d7d5 4. g1f3 f8d6 5. e2e3 g8f6 6. b1d2 e8g8 7. c2c4 c7c6 8. f1d3 b8d7 9. e1g1 f6e4 10. a1c1 g7g5 11. h2h3 d8e8 12. d3e4 d5e4 13. f3g5 e8g6 14. h3h4 h7h6 15. g5h3 d7f6 16. f2f4 e4f3 17. d2f3 f6g4 18. d1e2 d6g3 19. h3f4 g6g7 20. d4d5 g7f7 21. d5e6 c8e6 22. f3e5 g4e5 23. b2e5 g8h7 24. h4h5 f8g8 25. e2f3 g3f4 26. e5f4 g8g4 27. g2g3 a8g8 28. c1c2 b7b5 29. c4b5 e6d5 30. f3d1 f7h5 31. c2h2 g4g3+ 32. f4g3 g8g3+ 33. g1f2 h5h2+ 34. f2e1 g3g2 35. d1d3 d5e4 36. d3d7+ h7g6 37. b5c6 g2e2+ 38. e1d1 e2a2 0-1`;
 
-xdescribe('Game history for en passant move should show valid captured piece', () => {
+describe('Game history for en passant move should show valid captured piece', () => {
```
Comment (bennyrubanov) [inline tests/enPassant.test.ts:6]: test looks good to me. can you add expects for unambiguous pieces

File: tests/queensidecastle.test.ts [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import { Chess } from '../cjsmin/src/chess';
 
-xdescribe('Queenside castling is detected', () => {
+describe('Queenside castling is detected', () => {
```
Comment (bennyrubanov) [inline tests/queensidecastle.test.ts:3]: where does the .filter function come from? chess.js?

**Review by bennyrubanov — APPROVED** [review]

File: src/index.ts [hunk 14]
```diff
@@ -16,53 +15,50 @@ import { FileReaderGame } from './types';
...
       console.log('number of games ingested: ', gameCounter);
     }
-    games.push(game);
+    const siteLink = game.metadata[1].match(/"(.*?)"/)[1];
 
-    // const siteLink = game.metadata[1].match(/"(.*?)"/)[1];
-    // console.log(`lichess link to game played: ${siteLink}`);
+    for (const metric of metrics) {
+      const historyGenerator = cjsmin.historyGenerator(game.moves);
+      metric.processGame(Array.from(historyGenerator), siteLink);
```
[the remark this answers is not in the document]
Comment (EllAchE) [reply src/index.ts:57]: it is. I think below you mention passing the metadata. We can do that instead and it's probably a better pattern

File: src/metrics/distances.ts [hunk 15]
```diff
@@ -1,190 +1,156 @@
...
+      }
 
-    for (const piece of Object.keys(distanceMap)) {
-      if (!totalDistanceMap[piece]) {
-        totalDistanceMap[piece] = 0;
+      // find the individual piece maxes
+      if (singleGameMap[uas].total > this.pieceMaxes.distance) {
+        this.pieceMaxes.uasArray = [uas as UASymbol];
+        this.pieceMaxes.distance = singleGameMap[uas].total;
+        // this.singleGameMaxPiece.link = undefined; // TODO: add game link support
```
[the remark this answers is not in the document]
Comment (EllAchE) [reply src/metrics/distances.ts:148]: now passing md

File: src/metrics/distances.ts (hunk 2 above)
Comment (EllAchE) [reply src/metrics/distances.ts:64]: deleted

File: src/metrics/distances.ts (hunk 3 above)
Comment (EllAchE) [reply src/metrics/distances.ts:13]: using this in all spots now

File: src/metrics/distances.ts (hunk 4 above)
Comment (EllAchE) [reply src/metrics/distances.ts:139]: there are some typing issues with indexed based on the object mapper keys that I didn't want to deal with, one of these should be throwing a compiler error but it's masked by the object indexing. I've switched to just one

File: src/metrics/moves.ts (hunk 5 above)
Comment (EllAchE) [reply src/metrics/moves.ts:23]: should be? Each entry is a move

Comment (EllAchE) [reply src/metrics/moves.ts:23]: look @ the typing

File: src/metrics/promotions.ts (hunk 6 above)
Comment (EllAchE) [reply src/metrics/promotions.ts:10]: not 3d array. Object with depth 2 mapping keys to promotion count. We want to know what paws promotes to what. And we probably want to know on what square.

Writing the data to a db might be more useful here actually... But we are too deep

File: src/metrics/promotions.ts (hunk 7 above)
Comment (EllAchE) [reply src/metrics/promotions.ts:39]: yes

File: src/types.ts (hunk 8 above)
Comment (EllAchE) [reply src/types.ts:3]: and other things now

File: src/types.ts (hunk 9 above)
Comment (EllAchE) [reply src/types.ts:9]: look up generic types

File: src/utils.ts (hunk 11 above)
Comment (EllAchE) [reply src/utils.ts:60]: I'm using it rn

Comment (EllAchE) [reply src/utils.ts:60]: when we need a map of squares --> UAS --> some tracking counts

File: tests/enPassant.test.ts (hunk 12 above)
Comment (EllAchE) [reply tests/enPassant.test.ts:6]: all of your comments are above the code that actually matters. PR comments show the lines you select, or the 4 above your comment

Comment (EllAchE) [reply tests/enPassant.test.ts:6]: I don't understand what you wan the test here to do

File: tests/queensidecastle.test.ts (hunk 13 above)
Comment (EllAchE) [reply tests/queensidecastle.test.ts:3]: this is old test. Filter is a method applicable to arrays

**Pull request merged**